### ISSUE TITLE
Comparing Paths throws PHP Notice Error

### ISSUE THREAD

**Issue by k10blogger** [websvnphp/websvn#102]
I get the following warning when i try to compare any two paths. I am using svn 1.10.3, Apache version 2.4.35 with a SVN client 1.9.5

![image](https://user-images.githubusercontent.com/4973182/82129468-b7c79400-97e0-11ea-8f67-89658e10794c.png)

![image](https://user-images.githubusercontent.com/4973182/82129482-d75ebc80-97e0-11ea-8208-56c25415c297.png)

Can anybody tell me how to resolve this?

**Comment by ams-tschoening** [comment]
There's something wrong with the available data to WebSVN when comparing your paths. You need to debug this further in `com.php` around line 143.

	// Set variables used for the more recent of the two revisions
	$history = ($rev1 >= $rev2 ? $history1 : $history2);
	if ($history) {

to

	// Set variables used for the more recent of the two revisions
	$history = ($rev1 >= $rev2 ? $history1 : $history2);
	var_dump($history);
	var_dump($history1);
	var_dump($history2);
	if ($history) {

https://www.php.net/manual/en/function.var-dump.php

Results should be available in your web server logs. Didn't try it myself, might need some adjusting.

**Comment by k10blogger** [comment]
the version for WebSVN is 2.5.2-Dev

![Websvn version](https://user-images.githubusercontent.com/4973182/82149871-fa3ab080-9874-11ea-8a97-95dacf372b7d.png)

I added the 'var_dump()' lines as shown in lines numbers in the snapshot below:
![var_dump lines added](https://user-images.githubusercontent.com/4973182/82150301-cdd36400-9875-11ea-88ba-825ac7fa208e.png)

The following is what i get on the main area.
1. $history
2. $history1
3. $history2

![image](https://user-images.githubusercontent.com/4973182/82150349-17bc4a00-9876-11ea-98b0-2cf80ca9360c.png)

Following is the expansion of the dump. 
```
C:\wamp64\www\websvn\comp.php:143:
object(SVNLog)[5]
  public 'entries' => 
    array (size=0)
      empty
  public 'curEntry' => null
  public 'path' => string '/tags/gaLibTestScripts_0.6.0.1_01_12_2019' (length=41)
C:\wamp64\www\websvn\comp.php:144:
object(SVNLog)[5]
  public 'entries' => 
    array (size=0)
      empty
  public 'curEntry' => null
  public 'path' => string '/tags/gaLibTestScripts_0.6.0.1_01_12_2019' (length=41)
C:\wamp64\www\websvn\comp.php:145:
object(SVNLog)[6]
  public 'entries' => 
    array (size=0)
      empty
  public 'curEntry' => null
  public 'path' => string '/tags/gaLibTestScripts_1.0.0.99_27_04_2020' (length=42)
```



I dont see anything in the web server log except the php notices. But i guess the variable dump on the web page is equivalent.
![image](https://user-images.githubusercontent.com/4973182/82150805-c06aa980-9876-11ea-862a-35e14e03174f.png)

For access to SVN. I use a VisualSVNManager. I have a group called Tools that have Read Only access to all the repositories. I am pretty sure for comparison a read only access should do. But let me try by changing the access and get back to you. 
![Tools Group](https://user-images.githubusercontent.com/4973182/82150961-9fef1f00-9877-11ea-9d82-821fcbd0bb1a.png)
![websvn dedicated account](https://user-images.githubusercontent.com/4973182/82150970-b006fe80-9877-11ea-9ebd-a6d7dddfc6ff.png)

**Comment by ams-tschoening** [comment]
> For access to SVN. I use a VisualSVNManager. I have a group called Tools that have Read Only access to all the repositories. I am pretty sure for comparison a read only access should do.

Are you able to browse contents of the paths used during comparing at all? Current versions of WebSVN use `svnauthz accessof` to check access and you should check if this is compatible with you VisualSVN works at all.

Besides that, I'm afraid nothing obvious comes into my mind anymore, so you need to debug this further on your own. I tested comparing two paths in my version of WebSVN and that seemed to work with plain SVN.

**Comment by michael-o** [comment]
You should use the browser built in in VisualSVN because it does not use mod_dav_svn for authz in a way the default approach is not compatible with. VisualSVN extracts the groups SID from the PAC data embedded in the SPNEGO (Kerberos) ticket. Subversion does not have this kind of access to the Windows groups. Please recreate this case with pure HTTPd and Subversion provided modules.

**Comment by k10blogger** [comment]
Yeah i am able to browse the contents with the same id. if you see in the first snapshot in the WebSVN i am getting the comparison even though i am getting the warning.

I have another observation If i do the following:
Access with the last updated revision number of each tag or branch.
![image](https://user-images.githubusercontent.com/4973182/82156348-4cd99400-9898-11ea-8afc-022deee9c09d.png)

I dont get an error.

But if i do the following two i do get an error:
Access with the HEAD revision not a specific number
![image](https://user-images.githubusercontent.com/4973182/82156365-61b62780-9898-11ea-8f63-b8aae65c2486.png)

Access with the last revision number not tag or branch specific
![image](https://user-images.githubusercontent.com/4973182/82156399-827e7d00-9898-11ea-9e0e-d9fcdd3cbc8c.png)

Would this give you any hint??

**Comment by k10blogger** [comment]
@michael-o i did a little reading for you said. I am not using windows authentication. I am only using basic authentication. I am pretty sure that the Kerberos tokens are used, only when integrated windows authentication is used. That is not being user here. This is only plain basic authentication. If that is the scenario i should not get any comparisons at all even at different revision. I get comparisons without error when i put in the correct last revision of the path in the revision field.. As shown in the image in the comment above.

**Comment by michael-o** [comment]
@k10blogger Even if, we don't know how VisualSVN handles authorization since this is a closed source, commercial product. Can you recreate the same example with plain tools available from Subversion directly?

**Comment by ams-tschoening** [comment]
@k10blogger 

If `HEAD` vs. not makes such a difference in your case, try to do on the shell what WebSVN does [internally](https://github.com/websvnphp/websvn/blob/master/include/svnlook.php#L1054) to access the history:

    $cmd = $this->svnCommandString('log --xml '.($verbose ? '--verbose' : ($quiet ? '--quiet' : '')).($limit != 0 ? ' --limit '.$limit : ''), $path, $effectiveRev, $effectivePeg);

I would expect that there's a difference on the shell already and that is the root cause of this problem when using VisualSVN.

@michael-o 

Looking at the provided debugging output, I suggest the following changes:

	// Set variables used for the more recent of the two revisions
	$history = ($rev1 >= $rev2 ? $history1 : $history2);
	if ($history) {
		$logEntry = $history->curEntry;
                [...]
	} else {
		$vars['warning'] = 'Problem with comparison.';
	}

vs.

	// Set variables used for the more recent of the two revisions
	$history = ($rev1 >= $rev2 ? $history1 : $history2);
	if ($history && $history->curEntry) {
		$logEntry = $history->curEntry;
                [...]
	} else {
		$vars['warning'] = 'Problem with comparison.';
	}

This seems to me what would be expected in case of any problems and prevents ugly PHP warnings in the UI. Opinions?

**Comment by k10blogger** [comment]
@ams-tschoening it took me some time. I could not configured subversion on a local server. But i used the sourceforge subversion server running on The CentOS 7 distribution of Linux with Apache 2.4.x web server with Native Subversion. I was successfully able to recreate the issue there as well. so it is coming on the native subversion as well. 

![image](https://user-images.githubusercontent.com/4973182/82226488-2e30d700-9944-11ea-8d56-97b740cb7fad.png)

**Comment by k10blogger** [comment]
@ams-tschoening May i suggest the following change:
Instead of taking the head revision when "compare paths" button is hit why not take the revision number as highlighted that will solve the problem without having to suppress the error.
![image](https://user-images.githubusercontent.com/4973182/82227672-bd8aba00-9945-11ea-93bf-56e29b6cc416.png)

I wrote the commands written by Websvn have a look at below:
If i write with correct version number i do get an xml output but if i use the HEAD revision i do not get the xml output but no data. This is valid for both sourceforge one as well as the visualSVN one.

![image](https://user-images.githubusercontent.com/4973182/82227962-16f2e900-9946-11ea-8d69-2f94a8ce179f.png)

**Comment by ams-tschoening** [comment]
There are two problems here: Showing the `non-object` PHP-notices is wrong and should be handled properly. That's the code change I suggested. The underlying root cause is that HEAD is not working properly in your setup and because that's the case, the `non-object`-thing is wrongly triggered at all.

`HEAD` is supported in general by SVN, it's even documented that way:

	Valid options:
	  -r [--revision] ARG      : ARG (some commands also take ARG1:ARG2 range)
	                             A revision argument can be one of:
	                                NUMBER       revision number
	                                '{' DATE '}' revision at start of the date
	                                'HEAD'       latest in repository
	                                'BASE'       base rev of item's working copy
	                                'COMMITTED'  last commit at or before BASE
	                                'PREV'       revision just before COMMITTED

So the question is why it doesn't work as expected in your setup, while it does in my WebSVN-setup. OTOH, it doesn't work as expected on my shell as well:

	C:\[...]>svn log -r HEAD C:\[...]\trunk --xml --limit 1
	<?xml version="1.0" encoding="UTF-8"?>
	<log>
	</log>

vs. 

	C:\[...]>svn log -r COMMITTED C:\[...]\trunk --xml --limit 1
	<?xml version="1.0" encoding="UTF-8"?>
	<log>
	<logentry
	   revision="667">
	<author>tschoening@am-soft.de</author>
	<date>2020-04-04T08:11:58.036410Z</date>
	<msg>Tagged coming_soon to fit to the known workflow and added XXX as templates where in index.html things needs to be customized.</msg>
	</logentry>
	</log>

That needs to be investigated further before we can decide if to change using `HEAD` vs. concrete revision numbers etc.

**Comment by ams-tschoening** [comment]
There's an answer:

> rHEAD might have touched a path outside of trunk, or paths under trunk
that are authz-restricted, or it might actually be an empty revision.
To rule out these, look what paths rHEAD changed.  Use «svnlook changed»
if you can, or «svn log -qvrHEAD ^/» (double the caret if using cmd.exe
or zsh -o extendedglob).

> Another possibility is that trunk had been replaced between BASE and
HEAD.  Try passing HEAD as a peg revision.

> If none of these works, run «svn info -r HEAD --show-item=revision ^/»
before _and after_ reproducing the issue, and then see if you can
reproduce it against the specific revision number of HEAD at that time,
and we can take it from there.

https://mail-archives.apache.org/mod_mbox/subversion-users/202005.mbox/%3C1195409744.20200518182014%40am-soft.de%3E

**Comment by k10blogger** [comment]
We are trying to compare the tags and branches which are outside the trunk. 
The HEAD commit is not blank. 
The HEAD commit is on the trunk as seen from the below snapshot.
Also you can see the HEAD revision on a tag returns nothing whereas on the trunk or whole repository returns the commit.
![image](https://user-images.githubusercontent.com/4973182/82246186-d7d29100-9961-11ea-9222-3088962f3d1c.png)

Also from my understanding of the answer. I tried reproducing the issue with the specific revision number of HEAD. In my case 583. I still get the same error.
![image](https://user-images.githubusercontent.com/4973182/82246310-123c2e00-9962-11ea-8fc6-c9fa85ba553a.png)

Anything else i should look into to help?

**Comment by ams-tschoening** [comment]
There's another response on the list related to this problem:

> You first example looks exactly for the HEAD revision of the
repository. If you path was not touched by this revision, there is no
log regarding this path and revision to return.

> Your second example on the other hand explicitly asks for the log
message of the last revision the given path was touched.

This means `HEAD` is expected to sometimes NOT return any log history and my proposed code change will make things more compatible with that. Will provide a PR afterwards.

Knowing that, the second question is if to use `HEAD` without any explicit revision number is a good idea at all. I tried `COMMITTED` instead, because that always provides a revision number for a path and comparing somewhat succeeded. The problem is that this value CAN'T be used everywhere, one needs a working copy, while WebSVN executes some things using an URL only, like accessing the logs.

    C:\Users\tschoening>"C:/Program Files/CollabNet/Subversion Client\svn" --non-interactive --config-dir /tmp/websvn log --xml  --limit 1 -r COMMITTED:COMMITTED "file:///C:/[...]/trunk/keywords.txt@COMMITTED"
    <?xml version="1.0" encoding="UTF-8"?>
    <log>
    svn: E195002: PREV, BASE, or COMMITTED revision keywords are invalid for URL

`comp.php` doesn't support using different revisions for different operations currently.

This leaves us with not using keywords at all, but concrete revision numbers only. Where do those come from? Things depend on what is currently shown to the user, templates and stuff like that. The default is that revision number CAN'T be provided at all, but are hard-coded in the path to compare already:

    <input type="checkbox" name="compare[]" value="/branches/@0" onclick="enforceOnlyTwoChecked(this)">

`@0` is the thing used by `comp.php` and afterwards mapped to `HEAD`. That whole line gets build in `listing.php`, which in theory has the youngest revision number available for some path and in theory could be hard-coded into that value.

The problem is that showing directory contents can have an individual revision number as well and that is the one encoded in the above value. So things might look like this as well:

    <input type="checkbox" name="compare[]" value="/branches/@123" onclick="enforceOnlyTwoChecked(this)">

![rev123](https://user-images.githubusercontent.com/6223655/82297609-2e1bf000-99b3-11ea-9b93-4fc6c38297a0.png)

vs.

![rev0](https://user-images.githubusercontent.com/6223655/82297615-32480d80-99b3-11ea-9cc0-db6b11c037e8.png)

So not sure how to deal with this `HEAD` vs. `COMMITTED` vs. concrete revision number currently and am only removing the PHP-error notices for now.

**Comment by ams-tschoening** [comment]
@k10blogger Please have a look at PR #103, that should remove the PHP-errors. If that works for you, this issue will be closed. If results of comparing are not as expected, that should become a new issue based on my former comment.

**Comment by k10blogger** [comment]
@ams-tschoening This PR #103 works fine for me. I dont see any php-errors.

**Comment by michael-o** [comment]
> 
> 
> There's an answer:
> 
> > rHEAD might have touched a path outside of trunk, or paths under trunk
> > that are authz-restricted, or it might actually be an empty revision.
> > To rule out these, look what paths rHEAD changed.  Use «svnlook changed»
> > if you can, or «svn log -qvrHEAD ^/» (double the caret if using cmd.exe
> > or zsh -o extendedglob).
> 
> > Another possibility is that trunk had been replaced between BASE and
> > HEAD.  Try passing HEAD as a peg revision.
> 
> > If none of these works, run «svn info -r HEAD --show-item=revision ^/»
> > before _and after_ reproducing the issue, and then see if you can
> > reproduce it against the specific revision number of HEAD at that time,
> > and we can take it from there.
> 
> https://mail-archives.apache.org/mod_mbox/subversion-users/202005.mbox/%3C1195409744.20200518182014%40am-soft.de%3E

This sounds like a *lot* of hassle and work for something trivial.
